File: Data Structures & Algorithms/trapping-rain-water/submission-1.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height)-1
        prefix, suffix = [], []
        max_p, max_s = 0, 0

        for _ in range(len(height)):
            max_p = max(height[l], max_p)
            prefix.append(max_p)
            l+=1

            max_s = max(height[r], max_s)
            suffix.append(max_s)
            r-=1
        suffix = suffix[::-1]

        water = 0
        for i in range(len(height)):
            water += min(prefix[i], suffix[i]) - height[i]
        return water

File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0;
        int r = heights.size()-1;
        int maxArea = 0;
        while (l<r) {
            maxArea = max(maxArea, (r-l)*min(heights[r],heights[l]));
            if (heights[l] < heights[r]) {
                l++;
            } else {
                r--;
            }
        }
        return maxArea;
    }
};
